pkg/chunk/util/util.go

```diff
@@ -125,3 +127,21 @@ func EnsureDirectory(dir string) error {
 	}
 	return err
 }
+
+// ReadCloserWithContextCancelFunc helps with cancelling the context when closing a ReadCloser.
+type ReadCloserWithContextCancelFunc struct {
+	io.ReadCloser
+	cancel context.CancelFunc
+}
+
+func NewReadCloserWithContextCancelFunc(readCloser io.ReadCloser, cancel context.CancelFunc) io.ReadCloser {
```

Comment:
Should we also override the `Read()` function and cancel the context once `EOF` is reached? I'm not sure about the `ReadCloser` contract (if calling `Close()` is mandatory or not) and I can't find anything to clarify that in the golang doc. If a caller doesn't call `Close()` once the `EOF` is reached, we'll leak the memory. On the contrary, if we cancel once `EOF` (or maybe even error) is returned by `Close()` and we set `r.cancel = nil` (adding a check in `r.Close()` too), we'll protect from this case.

Thoughts?

Reply:
I think we should specify in `GetObject` contract that returned `ReadCloser` must be closed in order to avoid leaking resources. That's usually implied, but comment doesn't hurt. I don't think we need `Read` react on `EOF` in such case.

Reply:
Ok documenting it, but we should also check callers in Cortex to make sure they do it, otherwise we introduce an issue.